src/main/java/com/vmware/vipclient/i18n/messages/api/opt/local/LocalProductOpt.java

```diff
@@ -0,0 +1,113 @@
+import java.util.stream.Stream;
+
+public class LocalProductOpt implements ProductOpt {
+    private Logger logger = LoggerFactory.getLogger(LocalProductOpt.class);
+    private static final String BUNDLE_PREFIX = "messages_";
+    private static final String BUNDLE_SUFFIX = ".json";
+
+    private BaseDTO dto = null;
+
+    public LocalProductOpt(BaseDTO dto) {
+        this.dto = dto;
+    }
+
+    public List<String> getSupportedLocales() {
+        List<String> supportedLocales = new ArrayList<String>();
+        try {
+
+            Path path = Paths.get(VIPCfg.getInstance().getOfflineResourcesBaseUrl());
+
+            URI uri = Thread.currentThread().getContextClassLoader().
+                    getResource(path.toString()).toURI();
+
+            if (uri.getScheme().equals("jar")) {
+                try (FileSystem fileSystem = FileSystems.newFileSystem(uri, Collections.<String, Object>emptyMap())) {
+                    path = fileSystem.getPath(path.toString());
+                    getSupportedLocales(path, supportedLocales);
+                }
+            } else {
+                path = Paths.get(uri);
+                getSupportedLocales(path, supportedLocales);
+            }
+
+        } catch (Exception e) {
+            logger.debug(e.getMessage());
+        }
+        return supportedLocales;
+    }
+
+    public List<String> getComponents() {
+        List<String> components = new ArrayList<String>();
+        try {
+
+            Path path = Paths.get(VIPCfg.getInstance().getOfflineResourcesBaseUrl());
+
+            URI uri = Thread.currentThread().getContextClassLoader().
+                    getResource(path.toString()).toURI();
+
+            if (uri.getScheme().equals("jar")) {
+                try (FileSystem fileSystem = FileSystems.newFileSystem(uri, Collections.<String, Object>emptyMap())) {
+                    path = fileSystem.getPath(path.toString());
+                    getComponents(path, components);
+                }
+            } else {
+                path = Paths.get(uri);
+                getComponents(path, components);
+            }
+
+        } catch (Exception e) {
+            logger.debug(e.getMessage());
+        }
+        return components;
+    }
+
+    private void getSupportedLocales(Path path, List<String> supportedLocales) throws IOException {
+        try (Stream<Path> listOfFiles = Files.walk(path).filter(p -> Files.isRegularFile(p))) {
+            listOfFiles.map(file -> {
+                String fileName = file.getFileName().toString();
+                if(fileName.startsWith(BUNDLE_PREFIX) && fileName.endsWith(BUNDLE_SUFFIX)) {
+                    return fileName.substring(BUNDLE_PREFIX.length(), fileName.indexOf('.'));
+                }
+                return "";
+            }).forEach(language -> {
+                if(language != null && !language.isEmpty()) {
+                    supportedLocales.add(language);
+                }
+            });
+        }
+    }
```

Comment:
```suggestion
    private void getSupportedLocales(Path path, List<String> supportedLocales) throws IOException {
        try (Stream<Path> listOfFiles = Files.walk(path).filter(p -> Files.isRegularFile(p) && p.getFileName().startsWith(BUNDLE_PREFIX) && p.getFileName().endsWith(BUNDLE_SUFFIX))) {
            listOfFiles.map(file -> {
                return file.getFileName().toString().substring(BUNDLE_PREFIX.length(), fileName.indexOf('.'));
            }).forEach(language -> {
                supportedLocales.add(language);
            });
        }
    }
```

Reply:
These two conditions are for filtering file name, not for filter path,
fileName.getFileName().startsWith(BUNDLE_PREFIX) && fileName.getFileName().endsWith(BUNDLE_SUFFIX)

or locale list will be empty because no path meet these conditions.

Reply:
I made a mistake of not adding .toString() to p.getFileName(). I have updated the code suggestion.

Files.walk will iterate over ALL paths -- all folders AND files inside the root path and  the root path itself. So yes, it will apply to files as well.  The .filter part will check if each item is a file (not a folder) that starts with BUNDLE_PREFIX and ends with BUNDLE_SUFFIX. If one of these conditions fail, the item will not be included in listOfFiles. This way, you do not need the if statement. 

You also do not need to return an empty String so that you can remove `if(language != null && !language.isEmpty())`